MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/TrajectoryFilterWidget.py

```diff
@@ -0,0 +1,1463 @@
+
+    def set_snap(self, snap_to: float) -> None:
+        """Set the value to be snapped to modulo zero when spinbox value or text changes.
+
+        snap_to : float
+            Value to snap to.
+        """
+        self.reset_connections()
+        self.constraint = snap_to
+        callback = self.snap_to_value
+        self.valueChanged.connect(callback)
+        self.textChanged.connect(callback)
+
+    def setValue(self, val) -> None:
+        """Overrides setValue method of QDoubleSpinBox.
+        Sets a record of the initial value of the spinbox, for determination of direction of change.
+
+        Parameters
+        ----------
+        val : float
+            The value of the spinbox.
+        """
+        self.initial_value = val
+        super().setValue(val)
+
+    def snap_to_value(self, value: Any) -> None:
+        """Apply the constraint by snapping up/down (depending on the change direction) to the nearest value module zero.
+
+        Parameters
+        ----------
+        value : Any
+            The value of the spinbox.
+        """
+        value = self.to_float(value)
+
+        remainder = value % self.constraint
+        if not self.constraint or np.isclose(remainder, 0):
+            return
+
+        # New value is the closest value evenly dividing the constraint
+        new_value = np.round(value / self.constraint) * self.constraint
+        self.setValue(np.round(new_value, FLOAT_SPINBOX_DECIMALS))
+
+    def search_by_function(self, value: Any) -> None:
+        """Apply the constraint formalised in the lambda
+
+        Parameters
+        ----------
+        value : Any
+            The value of the spinbox.
+        """
+        value = self.to_float(value)
+
+        if (
+            not self.constraint
+            or self.constraint(value)
+            or (not hasattr(self, "initial_value"))
+        ):
+            return
+
+        current = self.to_float(self.value())
+        value_found = False
+        if current < self.initial_value:
+            for x in np.arange(value, self.minimum(), -self.singleStep()):
+                if self.constraint(x):
+                    value_found = True
+                    break
+        else:
+            for x in np.arange(value, self.maximum(), self.singleStep()):
+                if self.constraint(x):
+                    value_found = True
+                    break
+
+        if not value_found:
+            self.setValue(self.initial_value)
+            return
+
+        # Update with constrained value
+        self.setValue(x)
+
+    @staticmethod
+    def to_float(value: Any) -> float:
+        """Convert spinbox value to float if string
+
+        Parameters
+        ----------
+        value : Any
+            The value of the spinbox.
+
+        Returns
+        ----------
+        float
+            The value of the spinbox.
+        """
+        if isinstance(value, str):
+            return float(value)
+        return value
```

Comment:
```suggestion
        return float(value)
```
Why bother checking? Duck typing will take care of it if its invalid.